### PR TITLE
Implementation for CallbackConnection.h

### LINKED ISSUE #109: Consider switching to C++20
We have received feedback that many consumers of this library have already migrated to C++20. We should evaluate if it is safe to roll forward so that newer features can be used in the interface and implementation.

### LINKED ISSUE #112: Implement and test utils::callbacks
@gregmedd already has most of the implementation, so it just needs to be copied into the appropriate files and tested.

### PR DESCRIPTION
**Pull request by gregmedd**
Provides both the implementation and a test for the callback connection utility.

~~Posted in draft status until some accompanying documentation describing the design for this module can be added to the PR.~~

~~Note that the current change would switch the repo to C++20, pending confirmation that we are safe to do so. This is in part because the code in this PR heavily relies on C++20 features. While it is possible to rework to use C++17 features instead, some aspects of the implementation will be more complicated.~~

resolves #112

### CODE REVIEW

File: test/coverage/utils/CallbackConnectionTest.cpp
Comment (gregmedd) [inline test/coverage/utils/CallbackConnectionTest.cpp:1]: This test could really use some comments explaining what it is doing and why it's running specific checks.

File: include/up-cpp/utils/CallbackConnection.h
Comment (gregmedd) [inline include/up-cpp/utils/CallbackConnection.h:1]: Add a design doc file that at least explains the locking behavior around `sever()` with a sequence diagram.

File: CMakeLists.txt
Comment (gregmedd) [inline CMakeLists.txt:25]: Should the changes to switch to C++20 move to a different PR? Or are we OK reviewing them here since the implementation of the callback utils requires it right now.

**Comment by gregmedd** [comment]
The CI issues here are caused by a recent change in up-spec. A PR to fix it is pending: https://github.com/eclipse-uprotocol/up-spec/pull/161

File: .github/workflows/ci_conan_profile
Comment (stevenhartley) [inline .github/workflows/ci_conan_profile:1]: Why do we need this? Is this how CCI handles testing building for different configurations? How would we test for the arm/arch64, etc...

Comment (gregmedd) [reply .github/workflows/ci_conan_profile:1]: We need this to select C++20 instead of the default C++17 in the github build environment.

Profiles are how CCI (and conan in general) handle different build configurations and targets. We could add more profiles to test cross-compilation if we wanted. I believe github also has armv8 runners, so we could take the cross-compiled build then run the tests on a real arm core.

File: include/up-cpp/utils/CallbackConnection.h
Comment (gregmedd) [inline include/up-cpp/utils/CallbackConnection.h:245]: Unit test doesn't cover the true branch of this check.

**Comment by gregmedd** [comment]
Pulling back to draft status so I can split the commits.

File: CMakeLists.txt
Comment (gregmedd) [reply CMakeLists.txt]: The change to C++20 has been removed from this PR. New PRs to follow.

**Comment by gregmedd** [comment]
PR is ready for review again. C++20 changes have been split out and will be put in a separate PR.

**Review by debruce — COMMENTED** [review]
Given that there is a unit test, I would presume that technical validation of this design has occurred. Of course I will still go through it on a technical basis to review. However, up front, it seems reasonable to ask why this is needed at all, since it isn't a pattern I'm aware of being discussed for std inclusion, nor have I seen the idea in Boost. I did a working demo that does have the need for this.
It seems that registerListener() is the source of the need. The need for callbacks occur in rpc server, subscriber, and notification server, but not in rpc client since that is being handled by a future. In each needed case, there is deregistration on the Zenoh-cpp API or its alternative, and deregistration from possible thread pools or destruction of a thread pool, etc. It seems that the intent of this design is that registerListener() interacts with Zenoh-cpp, and the destruction of the thread pool or other resources has to be handled has a call path from higher level code. If the design were to have a subscriber object, for example, that implements Zenoh-cpp reg/dereg, and the thread pool, and allows for derivation or shared_ptr holding to bond the callback behavior to the subscriber object, then there is no possibility for things to get out-of-sync between listener registration and other resource deallocation.
Having said that, I think it's instructive to look at a hypothetical case of resources to be deallocated, because the simple callback case (for subscriber example) can just be a std::function<void (UMessage)>. A unit test for this feature should only have to deal with such a signature. Beyond a simple unit test, you may want to have the callback be able to record information that its seeing into a data structure for aggregate analysis on exit, or whatever. For a case like this, the top level object contains the aggregating info data structure, the callback as a callable method within this class (or a lambda function customizing to the class method), and a unique_ptr or suitable ref to the subscriber object. The parent object's life/death cycle contains the subscriber object's life/death cycle, and the subscriber object's life/death cycle controls Zenoh-cpp (or other) registration/deregistration under the hood. There should not be a registerListener() method. Its roile should be implicitly filled by the construction and destruction of the subscriber API object.
Having said this, I wish I could have gotten these basic points across a month ago.

**Comment by debruce** [comment]
In a follow-up to the above comment, I would repeat Steven's question about 'why do we need this'. Expecting C++20 to cover for a basic design issue seems risky and unwise. RAII suuggests a simpler path than this compositional approach extending up to the API.

File: docs/design/utils/CallbackConnection.md
Comment (gregmedd) [inline docs/design/utils/CallbackConnection.md:1]: Remove references to C++20 from this design. That can be saved for a future PR.

File: .github/workflows/ci_conan_profile
Comment (gregmedd) [reply .github/workflows/ci_conan_profile:1]: This has been removed. It will be handled in a separate PR.

**Comment by gregmedd** [comment]
> In a follow-up to the above comment, I would repeat Steven's question about 'why do we need this'.

@debruce - I am assuming you are talking about [this comment](https://github.com/eclipse-uprotocol/up-cpp/pull/123#discussion_r1613766594). Steven was asking why I had added a conan profile file to the CI. However, that has since been removed and this PR no longer involves any C++20 features.

> Expecting C++20 to cover for a basic design issue seems risky and unwise. RAII suuggests a simpler path than this compositional approach extending up to the API.

I don't know where you're coming from on this. I don't expect C++20 to _cover for a basic design issue_. It simply provides library features that implement the same functionality with less code on our end.

**Comment by gregmedd** [comment]
@debruce - I'm going to copy something I wrote near the end of my reply up to the top because I feel it might be the core of the misunderstanding here:

`Subscriber` is a layer 2 API and `UTransport` is a layer 1 API. It must be possible to use `UTransport` without using `Subscriber`. We have some use cases that will operate entirely at L1. We _cannot_ remove `registerListener()` from `UTransport`.

----

> However, up front, it seems reasonable to ask why this is needed at all, since it isn't a pattern I'm aware of being discussed for std inclusion, nor have I seen the idea in Boost.

It is needed to provide an alternative to a separate `unregisterListener()` interface. That risks leaving callbacks registered even after the code that initially registered the callback has long since returned and discarded its data. It is fairly common in the standard library to provide interfaces where the destruction of an object cleans up some state in order to avoid requiring a separate call to do so. See `std::lock_guard` and `std::unique_ptr`.

As for this exact concept being included in the std library or Boost: lots of code exists in the world that is not included in those places. It does not invalidate that other code. In the case of this particular model (representing a callback registration as a connection with discardable handles), I have seen it used very successfully in several code bases.

> I did a working demo that does have the need for this. It seems that registerListener() is the source of the need. The need for callbacks occur in rpc server, subscriber, and notification server, but not in rpc client since that is being handled by a future.

`registerListener()` and the issues caused by requiring a separate `unregisterListener()`, yes. That is the source of the need. Additionally, this provides a mechanism for the caller end to signal to the callee end that it is no longer using the callback (e.g. in the case of RPC response handlers, or when the transport has shut down).

> In each needed case, there is deregistration on the Zenoh-cpp API or its alternative, and deregistration from possible thread pools or destruction of a thread pool, etc.

I'm not sure what zenoh-cpp has to do with this at all.

> It seems that the intent of this design is that registerListener() interacts with Zenoh-cpp,

The intent of this design is that, when a callback is registered, a contract is formed between the calling and called parties. They are each given a handle representing that contract and, so long as that handle is held, the contract remains valid.

> and the destruction of the thread pool or other resources has to be handled has a call path from higher level code.

I don't follow.

> If the design were to have a subscriber object, for example, that implements Zenoh-cpp reg/dereg, and the thread pool, and allows for derivation or shared_ptr holding to bond the callback behavior to the subscriber object, then there is no possibility for things to get out-of-sync between listener registration and other resource deallocation.

TBH what you have described here does not sound too dissimilar from what is implemented here. However, you have to consider that the "subscriber" (callee) is potentially executing in a separate context from the transport, and callbacks could potentially be executing when the subscriber tries to deregister. Additional care is required to ensure that callbacks are allowed to return before the contract is broken.

> Having said that, I think it's instructive to look at a hypothetical case of resources to be deallocated, because the simple callback case (for subscriber example) can just be a std::function<void (UMessage)>.

Don't forget that the `std::function<void(UMessage)>` could be a lambda with captures.

> A unit test for this feature should only have to deal with such a signature.

Sure, but not all of the places we use this module have the exact same callback signature (see `RpcServer`). It adds very little actual complexity to make it generic.

> Beyond a simple unit test, you may want to have the callback be able to record information that its seeing into a data structure for aggregate analysis on exit, or whatever. For a case like this, the top level object contains the aggregating info data structure, the callback as a callable method within this class (or a lambda function customizing to the class method), and a unique_ptr or suitable ref to the subscriber object.

The callback _can_ access other data to record aggregate information. The callback can access class members if it is a class method, or it can access anything it captures if it is a lambda. Alternatively, interfaces wanting to allow callbacks to return a value (see `RpcServer`) can do so by setting the `RT` template parameter to some non-void value.

Providing a mechanism for callbacks to aggregate information goes beyond the scope of the code in this review.

> The parent object's life/death cycle contains the subscriber object's life/death cycle, and the subscriber object's life/death cycle controls Zenoh-cpp (or other) registration/deregistration under the hood. There should not be a registerListener() method. Its roile should be implicitly filled by the construction and destruction of the subscriber API object. Having said this, I wish I could have gotten these basic points across a month ago.

I feel like this has been said many times, but `Subscriber` is a L2 API and `UTransport` is a L1 API. It must be possible to use `UTransport` without using `Subscriber`. We have some use cases that will operate entirely at L1. We _cannot_ remove `registerListener()` from `UTransport`.

However, most uE code _will_ use L2 interfaces like `Subscriber`, and in that case the entire `registerListener()` thing is done for them by the L2 object.

**Comment by gregmedd** [comment]
@debruce - I have a request going forward: Can you please make all comments from the _files changed_ tab of this PR, highlighting the exact lines you are commenting on? I would like to keep this review focused on whether the code is ready to be merged. The use of this model for representing callbacks has already been litigated several times.

**Review by stevenhartley — APPROVED** [review]
LGTM, please get +1 from another C++ developer

File: include/up-cpp/utils/CallbackConnection.h
Comment (gregmedd) [inline include/up-cpp/utils/CallbackConnection.h:168]: So I just figured out a way to combine both `operator()` versions into a single implementation. Any interest in me adding that to this PR? I think it would probably be better in the long run; it avoids two places where possible fixes could be needed.

Comment (billpittman) [reply include/up-cpp/utils/CallbackConnection.h:168]: I suggest putting that change in another PR, this one is close to submission.

Comment (gregmedd) [reply include/up-cpp/utils/CallbackConnection.h:168]: I'll hold off on this unless there are other changes requested.

File: include/up-cpp/utils/CallbackConnection.h
Comment (MarkStokes71) [inline include/up-cpp/utils/CallbackConnection.h:363]: These uses of auto are non-compliant to Autosar.  Here's a reference: Rule A7-1-5

File: include/up-cpp/utils/CallbackConnection.h
Comment (MarkStokes71) [inline include/up-cpp/utils/CallbackConnection.h:383]: Is the point of this to allow sorted lists by caller address?  This doesn't seem to provide any value.

File: test/coverage/utils/CallbackConnectionTest.cpp
Comment (MarkStokes71) [inline test/coverage/utils/CallbackConnectionTest.cpp:145]: Point of note, test code does not need to be Autosar compliant.

**Review by MarkStokes71 — COMMENTED** [review]
Nice work.

File: include/up-cpp/utils/CallbackConnection.h
Comment (gregmedd) [reply include/up-cpp/utils/CallbackConnection.h:383]: It allows for handles to be stored in a `std::map` if needed for reverse lookups during the cleanup operation

File: include/up-cpp/utils/CallbackConnection.h
Comment (gregmedd) [reply include/up-cpp/utils/CallbackConnection.h:363]: Autosar is not, to my knowledge, a requirement of up-cpp

File: test/coverage/utils/CallbackConnectionTest.cpp
Comment (gregmedd) [reply test/coverage/utils/CallbackConnectionTest.cpp:145]: Per Autosar A7-1-5:

> The auto specifier shall not be used apart from following cases: (1) to declare
that a variable has the same type as return type of a function call...

This use of `auto` is Autosar compliant.

File: include/up-cpp/utils/CallbackConnection.h
Comment (gregmedd) [reply include/up-cpp/utils/CallbackConnection.h:363]: Reviewing Autosar A7-1-5, this use of `auto` would be allowed with a trailing return type to aid in type deduction:

> The auto specifier shall not be used apart from following cases: ... (4) to declare a function template using trailing return type syntax.

Based on that, a compliant fix would be:
```suggestion
	auto operator()(Args&&... args) -> decltype(*connection_)(std::forward<Args>(args)...)) {
```

(yes, the examples provided for a valid case 4 in Autosar A7-1-5 do use `decltype()` like that)

Personally, I find using a trailing return type in this context redundant - the exact same type deduction will occur regardless since the code that generates the return is copied twice.

Incidentally, Autosar has been merged with MISRA for C++17 and up based on what I've read, so the referenced rules are likely out of date for projects using anything after C++14.

File: include/up-cpp/utils/CallbackConnection.h
Comment (gregmedd) [reply include/up-cpp/utils/CallbackConnection.h:168]: Change committed. I'll push it once I've addressed more of the comments.

File: include/up-cpp/utils/CallbackConnection.h
Comment (gregmedd) [reply include/up-cpp/utils/CallbackConnection.h:100]: Added `isConnected()` methods.

File: include/up-cpp/utils/CallbackConnection.h
Comment (gregmedd) [reply include/up-cpp/utils/CallbackConnection.h]: Removed. It does mean that the `typename` keyword becomes necessary, so instead of `shared_ptr<Callback>` it will read `shared_ptr<typename Conn::Callback>`.

File: include/up-cpp/utils/CallbackConnection.h
Comment (gregmedd) [reply include/up-cpp/utils/CallbackConnection.h:239]: I agree - I don't like them. But it's not possible to have `using Connection = Connection<RT, Args...>` and putting the full templating list everywhere, in my opinion, makes the code harder to read. So given the list of not great choices, this isn't the worst option for readability.

File: test/coverage/utils/CallbackConnectionTest.cpp
Comment (gregmedd) [reply test/coverage/utils/CallbackConnectionTest.cpp:150]: (I know I submitted another comment here, but it's not showing up so I'm trying again)

I must have needed more sleep when I wrote that earlier comment. The letter of the requirement in A7-1-5 would say that `auto other_callable = callable` is not allowed. However, the _meaning_ of exception 1 effectively applies here. There is no intervening code between the line before where a value is assigned auto from a function return and where that value is immediately copied to another value. The type inferences operate the same in both cases. There would be little to no value in the first use of `auto` it the second use were not allowed.

File: include/up-cpp/utils/CallbackConnection.h
Comment (gregmedd) [reply include/up-cpp/utils/CallbackConnection.h:375]: I think where we are at is that we need to see if we can get our hands on MISRA C++ 2023 to check the latest guidance for C++17 and C++20. We will also be asking for static analysis to be run against this code base. For the moment, I'd like to defer these changes until we have the appropriate references. There are several P0 and P1 issues waiting on the completion of this module. We will make a clean-up pass when possible to address MISRA/AUTOSAR compliance.

File: test/coverage/utils/CallbackConnectionTest.cpp
Comment (gregmedd) [reply test/coverage/utils/CallbackConnectionTest.cpp:157]: Tracking request in #154

File: include/up-cpp/utils/CallbackConnection.h
Comment (gregmedd) [reply include/up-cpp/utils/CallbackConnection.h:375]: Tracking request in #154

**Comment by gregmedd** [comment]
All open threads have either been resolved or moved to an issue to track for future work. Once an approval is added (@MarkStokes71, if you are ok with the state of this PR), this PR can be merged.

**Review by MarkStokes71 — APPROVED** [review]

**Pull request merged**
